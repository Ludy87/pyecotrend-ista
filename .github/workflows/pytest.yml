name: Test

on:
  push:
    branches: [main]
    # paths-ignore:
    #   - ".github/**"
  pull_request:
    branches: [main]

permissions:
  contents: read

jobs:
  test:
    runs-on: ubuntu-latest
    strategy:
      matrix:
        python_version: ["3.11", "3.12", "3.13"]

    steps:
      - name: Harden Runner
        uses: step-security/harden-runner@0634a2670c59f64b4a01f0f96f84700a4088b9f0 # v2.12.0
        with:
          egress-policy: audit

      - uses: actions/checkout@11bd71901bbe5b1630ceea73d27597364c9af683 # v4.2.2
      - name: Set up Python
        uses: actions/setup-python@a26af69be951a213d495a4c3e4e4022e16d87065 # v5.6.0
        with:
          python-version: ${{ matrix.python_version }}
      - name: Install dependencies
        run: |
          python -m pip install --upgrade pip
          pip install .[test]
      - name: Run tests with tox
        run: tox -e py${{ matrix.python_version }}
      - uses: codecov/codecov-action@18283e04ce6e62d37312384ff67231eb8fd56d24 # v5.4.3
        with:
          token: ${{ secrets.CODECOV_TOKEN }}
          slug: Ludy87/pyecotrend-ista
          fail_ci_if_error: false
          verbose: true
