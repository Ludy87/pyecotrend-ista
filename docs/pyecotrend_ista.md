# API documentation

::: pyecotrend_ista.PyEcotrendIsta
    :docstring:
    :members:
<!--
::: pyecotrend_ista.openid
    :docstring:
    :members:

::: pyecotrend_ista.types
    :docstring:
    :members:

::: pyecotrend_ista.login_helper
    :docstring:
    :members:

::: pyecotrend_ista.exception_classes
    :docstring:
    :members: -->
