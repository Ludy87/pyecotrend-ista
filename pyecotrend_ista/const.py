VERSION = "1.0.33"
LOGIN_HEADER = {
    "Accept": "application/json, text/plain, */*",
    "Accept-Encoding": "gzip, deflate, br",
    "Accept-Language": "de-DE,de;q=0.9,en-US;q=0.8,en;q=0.7",
    "Connection": "keep-alive",
    "Content-Type": "application/json",
    "Origin": "https://ecotrend.ista.de",
    "Referer": "https://ecotrend.ista.de/",
    "Sec-Fetch-Dest": "empty",
    "Sec-Fetch-Mode": "cors",
    "Sec-Fetch-Site": "cross-site",
    "Sec-GPC": "1",
}
LOGIN_URL = "https://api.prod.eed.ista.com/login"
