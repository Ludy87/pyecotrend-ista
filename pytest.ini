[pytest]
addopts =  --cov=pyecotrend_ista --cov-report=term-missing
testpaths =
    tests

pythonpath =
    src
