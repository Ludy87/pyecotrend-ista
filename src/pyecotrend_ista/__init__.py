"""pyecotrend ista."""

from .exception_classes import (
    Error,
    InternalServerError,
    KeycloakAuthenticationError,
    KeycloakAuthorizationConfigError,
    KeycloakConnectionError,
    KeycloakDeleteError,
    KeycloakDeprecationError,
    KeycloakError,
    KeycloakGetError,
    KeycloakInvalidTokenError,
    KeycloakOperationError,
    KeycloakPostError,
    KeycloakPutError,
    KeycloakRPTNotFound,
    KeycloakSecretNotFound,
    LoginError,
    ServerError,
)

__all__ = [
    KeycloakAuthenticationError,
    KeycloakAuthorizationConfigError,
    KeycloakConnectionError,
    KeycloakDeleteError,
    KeycloakDeprecationError,
    KeycloakError,
    KeycloakGetError,
    KeycloakInvalidTokenError,
    KeycloakOperationError,
    KeycloakPostError,
    KeycloakPutError,
    KeycloakRPTNotFound,
    KeycloakSecretNotFound,
    LoginError,
    InternalServerError,
    ServerError,
    Error,
]
