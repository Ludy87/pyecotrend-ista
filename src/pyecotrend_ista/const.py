VERSION = "1.0.37"
LOGIN_HEADER = {
    "Accept": "application/json, text/plain, */*",
    "Accept-Encoding": "gzip, deflate, br",
    "Accept-Language": "de-DE,de;q=0.9,en-US;q=0.8,en;q=0.7,pt;q=0.6,zh-CN;q=0.5,zh;q=0.4,da;q=0.3",
    "Connection": "keep-alive",
    "Content-Type": "application/json",
    "Origin": "https://ecotrend.ista.de",
    "Referer": "https://ecotrend.ista.de/",
    "Sec-Fetch-Dest": "empty",
    "Sec-Fetch-Mode": "cors",
    "Sec-Fetch-Site": "cross-site",
    "Sec-GPC": "1",
}
LOGIN_URL = "https://api.prod.eed.ista.com/login"
