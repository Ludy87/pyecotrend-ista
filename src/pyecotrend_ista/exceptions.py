"""Exception Class."""

from __future__ import annotations

from typing import TypeVar

T = TypeVar("T")

class BaseError(Exception):
    """
    Base class for exceptions in this module.

    This is the base class for all custom exceptions in the module. It inherits
    from Python's built-in Exception class and can be used to catch errors specific
    to this module.
    """

class ServerError(BaseError):
    """Exception raised for server errors during requests.

    This exception is raised when a exception occurs during a request.
    It inherits from BaseError  and can be used to handle server-related
    issues specifically.
    """

    def __str__(self) -> str:
        """Return a string representation of the error.."""
        return "Server error occurred during the request"


class LoginError(BaseError):
    """Exception raised for login- and authentication related errors.

    This exception is raised when an authentication exception occurs during a request.
    It inherits from BaseError and is used specifically to handle issues related
    to authentication and login.
    """

    def __str__(self) -> str:
        """Return a string representation of an authentication error."""
        return "An authentication error occurred during the request"

class ParserError(ServerError):
    """Exception raised for errors encountered during parsing.

    This exception is raised when an error occurs during the parsing process
    of the request response. It inherits from BaseError and can be used
    to handle issues specifically related to parsing.
    """

    def __str__(self) -> str:
        """Return a string representation of parser error."""
        return "Error occurred during parsing of the request response"


# The MIT License (MIT)
#
# Copyright (C) 2017 Marcos Pereira <marcospereira.mpj@gmail.com>
#
# Permission is hereby granted, free of charge, to any person obtaining a copy of
# this software and associated documentation files (the "Software"), to deal in
# the Software without restriction, including without limitation the rights to
# use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of
# the Software, and to permit persons to whom the Software is furnished to do so,
# subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in all
# copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS
# FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR
# COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER
# IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
# CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.

# Source from https://github.com/marcospereirampj/python-keycloak/blob/c98189ca6951f12f1023ed3370c9aaa0d81e4aa4/src/keycloak/exceptions.py

# Keycloak custom exceptions module.


class KeycloakError(Exception):
    """Base class for custom Keycloak errors.

    :param error_message: The error message
    :type error_message: str
    :param response_code: The response status code
    :type response_code: int
    """

    def __init__(self, error_message="", response_code=None, response_body=None):
        """Init method.

        :param error_message: The error message
        :type error_message: str
        :param response_code: The code of the response
        :type response_code: int
        :param response_body: Body of the response
        :type response_body: bytes
        """
        Exception.__init__(self, error_message)

        self.response_code = response_code
        self.response_body = response_body
        self.error_message = error_message

    def __str__(self):
        """Str method.

        :returns: String representation of the object
        :rtype: str
        """
        if self.response_code is not None:
            return f"{self.response_code}: {self.error_message}"
        else:
            return f"{self.error_message}"


class KeycloakAuthenticationError(KeycloakError):
    """Keycloak authentication error exception."""


class KeycloakOperationError(KeycloakError):
    """Keycloak operation error exception."""


class KeycloakGetError(KeycloakOperationError):
    """Keycloak request get error exception."""


class KeycloakPostError(KeycloakOperationError):
    """Keycloak request post error exception."""


class KeycloakCodeNotFound(KeycloakOperationError):  # noqa: N818
    """Keycloak Code not found exception."""


class KeycloakInvalidTokenError(KeycloakOperationError):
    """Keycloak invalid token exception."""
