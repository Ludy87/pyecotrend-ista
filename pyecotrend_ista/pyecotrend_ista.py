from __future__ import annotations

from asyncio import sleep, TimeoutError
from datetime import datetime
from random import randint
from typing import Any, Dict, List

import aiohttp
import json
import logging

from .exception_classes import LoginError, ServerError

from .const import LOGIN_HEADER, LOGIN_URL, VERSION

_LOGGER = logging.getLogger(__name__)


class PyEcotrendIsta:
    def __init__(self, email: str, password: str) -> None:
        self._accessToken: str | None = None
        self._header: Dict[str, str] = {}
        self._supportCode: str | None = None
        self._uuid: str | None = None

        self._version = VERSION

        self.maxRetries = 3
        self.retryDelay = 2

        self._email = email
        self._password = password

        self._custom_types: List[str] = []

    def _isConnected(self) -> bool:
        if self._accessToken:
            return True
        return False

    def _logoff(self) -> None:
        self._accessToken = None

    async def __login(self) -> str | None:
        payload = {
            "email": self._email,
            "password": self._password,
            "fromMobileApp": "true",
        }
        LOGIN_HEADER["User-Agent"] = await self.getUA()
        async with aiohttp.ClientSession() as session:
            async with session.post(LOGIN_URL, headers=LOGIN_HEADER, data=json.dumps(payload)) as response:
                try:
                    if response.status == 500:
                        raise ServerError()
                    elif response.status != 200:
                        raise LoginError(await response.json())
                    else:
                        json_str_resp = await response.json()
                        self._accessToken = json_str_resp["accessToken"]
                except Exception as err:
                    _LOGGER.debug(err)
                    raise LoginError(err)
                finally:
                    await session.close()
        return self._accessToken

    async def __setAccount(self) -> None:
        self._header = LOGIN_HEADER
        del self._header["Accept-Encoding"]
        del self._header["Content-Type"]
        self._header["User-Agent"] = await self.getUA()
        self._header["Authorization"] = "Bearer {}".format(self._accessToken)
        async with aiohttp.ClientSession() as session:
            async with session.get("https://api.prod.eed.ista.com/account", headers=self._header) as response:
                res = await response.json()
                await session.close()
                self._a_ads = res["ads"]
                self._a_authcode = res["authcode"]
                self._a_betaPhase = res["betaPhase"]
                self._a_consumptionUnitUuids = res["consumptionUnitUuids"]
                self._a_country = res["country"]
                self._a_email = res["email"]
                self._a_emailConfirmed = res["emailConfirmed"]
                self._a_enabled = res["enabled"]
                self._a_fcmToken = res["fcmToken"]
                self._a_firstName = res["firstName"]
                self._a_isDemo = res["isDemo"]
                self._a_keycloakId = res["keycloakId"]
                self._a_lastName = res["lastName"]
                self._a_locale = res["locale"]
                self._a_marketing = res["marketing"]
                self._a_mobileLoginStatus = res["mobileLoginStatus"]
                self._a_notificationMethod = res["notificationMethod"]
                self._a_notificationMethodEmailConfirmed = res["notificationMethodEmailConfirmed"]
                self._a_password = res["password"]
                self._a_privacy = res["privacy"]
                self._a_residentAndConsumptionUuidsMap = res["residentAndConsumptionUuidsMap"]
                self._a_residentTimeRangeUuids = res["residentTimeRangeUuids"]
                self._supportCode = res["supportCode"]
                self._a_tos = res["tos"]
                self._a_tosUpdated = res["tosUpdated"]
                self._a_transitionMobileNumber = res["transitionMobileNumber"]
                self._a_unconfirmedPhoneNumber = res["unconfirmedPhoneNumber"]
                self._a_userGroup = res["userGroup"]
                self._uuid = res["activeConsumptionUnit"]

    def getVersion(self) -> str:
        return self._version

    async def login(self, forceLogin: bool = False) -> str | None:
        if not self._isConnected() or forceLogin:
            try:
                self._logoff()
                retryCounter = 0
                while not self._isConnected() and (retryCounter < self.maxRetries + 2):
                    retryCounter += 1
                    try:
                        self._accessToken = await self.__login()
                    except Exception as err:
                        raise Exception(err)
                    if not self._accessToken:
                        await sleep(self.retryDelay)
                    else:
                        await self.__setAccount()
            except LoginError as err:
                # Login failed
                self._accessToken = None
                _LOGGER.debug(err)
        return self._accessToken

    async def consum_raw(self) -> Dict[str, Any]:
        c_raw: Dict[str, Any] = {}
        timeout = aiohttp.ClientTimeout(total=12)
        async with aiohttp.ClientSession(timeout=timeout) as session:
            async with session.get(
                "https://api.prod.eed.ista.com/consumptions?consumptionUnitUuid={}".format(self._uuid),
                headers=self._header,
            ) as response:
                retryCounter = 0
                while not c_raw and (retryCounter < self.maxRetries + 2):
                    retryCounter += 1
                    try:
                        c_raw = await response.json()
                        if "key" in c_raw:
                            raise Exception("Login fail, check your input!", c_raw["key"])
                    except TimeoutError as error:
                        _LOGGER.debug(error)
                await session.close()
        return c_raw

    def getSupportCode(self) -> str | None:
        return self._supportCode

    async def consum_small(self) -> List[Dict[str, Any]]:
        consum_raw: Dict[str, Any] = {}
        retryCounter = 0
        _consum: List[Dict[str, Any]] = []
        while not consum_raw and (retryCounter < self.maxRetries + 2):
            retryCounter += 1
            await self.login()
            consum_raw = await self.consum_raw()
            if not consum_raw.get("consumptions", []):
                await sleep(self.retryDelay)
        if not consum_raw.get("consumptions", []):
            raise Exception("No Data found!")
        consumptions: List[Dict[str, Any]] = consum_raw.get("consumptions", [])
        for consum in consumptions:
            readings: List[Dict[str, Any]] = consum.get("readings", {})
            for reading in readings:
                if reading.get("type", ""):
                    self._custom_types.append(reading.get("type", ""))
                    _consum.append(
                        {
                            "entity_id": "{}_{}_{}_{}".format(
                                # sensor.warmwasser_yyyy_m_xxxxxxxxx
                                reading["type"],
                                consum["date"]["year"],
                                consum["date"]["month"],
                                str(self._supportCode).lower(),
                            ),
                            "year": consum["date"]["year"],
                            "month": consum["date"]["month"],
                            "type": reading["type"],
                            "value": reading["value"],
                            "valuekwh": reading["additionalValue"],
                            "unit": reading["unit"],
                            "unitkwh": reading["additionalUnit"],
                            "supportCode": self._supportCode,
                            "date": consum["date"],
                            "exception": consum["exception"],
                        }
                    )
        return _consum

    async def getTypes(self) -> List[str]:
        await self.consum_small()
        return list(dict.fromkeys(self._custom_types))

    async def getConsumsNow(self) -> List[Dict[str, Any]]:
        datetimenow = datetime.now()
        _consums: List[Dict[str, Any]] = []
        consums: List[Dict[str, Any]] = await self.consum_small()
        for consum in consums:
            consum["entity_id"] = "{}_{}".format(
                # sensor.warmwasser_xxxxxxxxx
                consum["type"],
                self._supportCode,
            ).lower()
            if datetimenow.year == consum.get("year", 0) and datetimenow.month == (consum.get("month", 0) + 1):
                _consums.append(consum)
        return _consums

    async def getConsumNowByType(self, _type: str | None) -> Dict[str, Any]:
        consums: List[Dict[str, Any]] = await self.consum_small()
        for consum in consums:
            if _type == consum["type"]:
                return consum
        return {}

    async def getConsumById(self, entity_id: str | None, now: bool = False) -> Dict[str, Any]:
        consums: List[Dict[str, Any]] = []
        if now is False:
            consums = await self.consum_small()
        else:
            consums = await self.getConsumsNow()
        for consum in consums:
            if entity_id == consum["entity_id"]:
                return consum
        return {}

    async def getConsumsByType(self, _type: str | None) -> List[Dict[str, Any]]:
        __type: List[Dict[str, Any]] = []
        consums: List[Dict[str, Any]] = await self.consum_small()
        for consum in consums:
            if _type == consum["type"]:
                __type.append(consum)
        return __type

    async def getConsumsByYear(self, year: int | None) -> List[Dict[str, Any]]:
        __type: List[Dict[str, Any]] = []
        consums: List[Dict[str, Any]] = await self.consum_small()
        for consum in consums:
            if year == consum["year"]:
                __type.append(consum)
        return __type

    async def getConsumsByMonth(self, month: int | None) -> List[Dict[str, Any]]:
        __type: List[Dict[str, Any]] = []
        consums: List[Dict[str, Any]] = await self.consum_small()
        for consum in consums:
            if month == consum["month"]:
                __type.append(consum)
        return __type

    async def getConsumsByYearMonth(self, month: int | None, year: int | None) -> List[Dict[str, Any]]:
        __type: List[Dict[str, Any]] = []
        consums: List[Dict[str, Any]] = await self.consum_small()
        for consum in consums:
            if month == consum["month"] and year == consum["year"]:
                __type.append(consum)
        return __type

    async def getUA(self) -> str:
        url = "https://raw.githubusercontent.com/Ludy87/pyecotrend-ista/main/pyecotrend_ista/ua.json"
        headers = {
            "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/99.0.4844.51 Safari/537.36"
        }
        _data = headers.get("User-Agent", "")
        async with aiohttp.ClientSession() as session:
            async with session.get(url, headers=headers) as response:
                data = await response.json(content_type=None)
                response.close()
                i = randint(0, len(data) - 1)
                _data = data[i]["useragent"]
        return _data
