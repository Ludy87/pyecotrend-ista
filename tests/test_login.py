"""Tests for Login methods."""

from unittest.mock import MagicMock

import pytest
import requests
from requests.exceptions import JSONDecodeError
from requests_mock.mocker import Mocker as RequestsMock
from syrupy.assertion import SnapshotAssertion

from pyecotrend_ista import PyEcotrendIsta, ServerError
from pyecotrend_ista.const import API_BASE_URL
from tests.conftest import DEMO_EMAIL, TEST_EMAIL


@pytest.mark.parametrize("ista_client", [TEST_EMAIL, DEMO_EMAIL], indirect=True)
@pytest.mark.usefixtures("mock_requests_login")
def test_login(ista_client: PyEcotrendIsta) -> None:
    """Test Login method."""
    assert ista_client.login() == "ACCESS_TOKEN"


@pytest.mark.parametrize("ista_client", [DEMO_EMAIL], indirect=True)
@pytest.mark.usefixtures("mock_requests_login")
def test_demo_user_login(ista_client: PyEcotrendIsta, snapshot: SnapshotAssertion) -> None:
    """Test Login method."""

    assert ista_client.demo_user_login() == snapshot



@pytest.mark.parametrize("ista_client", [DEMO_EMAIL], indirect=True)
@pytest.mark.parametrize(
    ("exception", "expected_exception"), ([(requests.RequestException, ServerError), (requests.Timeout, ServerError)])
)
def test_demo_user_login_exceptions(
    ista_client: PyEcotrendIsta,
    mock_requests_login: RequestsMock,
    exception,
    expected_exception,
) -> None:
    """Test exceptions for method `demo_user_login`."""

    mock_requests_login.get(f"{API_BASE_URL}demo-user-token", exc=exception)

    with pytest.raises(expected_exception=expected_exception):
        ista_client.demo_user_login()


@pytest.mark.parametrize("ista_client", [DEMO_EMAIL], indirect=True)
def test_demo_user_login_parser_exception(ista_client: PyEcotrendIsta, mock_requests_login: RequestsMock) -> None:
    """Test JSONDecodeError exception for method `demo_user_login`."""

    json_encoder = MagicMock().json.side_effect = JSONDecodeError("test", "test", 0)
    mock_requests_login.get(f"{API_BASE_URL}demo-user-token", json_encoder=json_encoder)

    with pytest.raises(expected_exception=ServerError):
        ista_client.demo_user_login()

