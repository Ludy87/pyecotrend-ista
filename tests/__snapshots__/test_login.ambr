# serializer version: 1
# name: test_demo_user_login[demo@ista.de]
  dict({
    'access_token': 'ACCESS_TOKEN',
    'expires_in': 60,
    'refresh_expires_in': 5184000,
    'refresh_token': 'REFRESH_TOKEN',
  })
# ---
